sync/class.jetpack-sync-actions.php

```diff
@@ -178,18 +182,58 @@ static function send_data( $data, $codec_name, $sent_timestamp, $queue_id ) {
 	}
 
 	static function do_initial_sync( $new_version = null, $old_version = null ) {
-		$initial_sync_config = array(
+		$initial_sync_config = self::get_update_full_sync_config();
+		$include_users = false;
+		if ( $old_version && ( version_compare( $old_version, '4.2', '<' ) ) ) {
+			$initial_sync_config['users'] = 'initial';
+			$include_users = true;
+		}
+
+		if ( is_multisite()  ) {
+			// stagger initial syncs for multisite blogs so they don't all pile on top of each other
+			if ( is_main_site() ) {
+				wp_schedule_single_event( time() , 'jetpack_sync_update_multisite_cron', $include_users );
+			}
+		} else {
+			self::do_full_sync( $initial_sync_config );
+		}
+	}
+
+	static function get_update_full_sync_config() {
+		return array(
 			'options' => true,
 			'network_options' => true,
 			'functions' => true,
 			'constants' => true,
 		);
-
-		if ( $old_version && ( version_compare( $old_version, '4.2', '<' ) ) ) {
+	}
+	
+	static function update_multisite_full_sync( $include_users = false ) {
+		global $wpdb;
+		$initial_sync_config = self::get_update_full_sync_config();
+		if ( $include_users ) {
 			$initial_sync_config['users'] = 'initial';
 		}
 
-		self::do_full_sync( $initial_sync_config );
+		$offset = 0;
+		$batch = 500;
+		$continue = true;
+
+		while( $continue ) {
+
+			$site_ids = $wpdb->get_col( "SELECT blog_id FROM {$wpdb->blogs} WHERE site_id = '{$wpdb->siteid}' AND spam = '0' AND deleted = '0' AND archived = '0' ORDER BY blog_id DESC LIMIT {$offset}, {$batch}" );
```

Comment:
On very large networks (millions of sites) this will get extremely slow as you paginate due to silly MySQL IO weirdness. Better to structure the query like this:

```php
$max_blog_id = 0;

while( true) {
  $site_ids = $wpdb->get_col( "SELECT blog_id FROM {$wpdb->blogs} WHERE site_id = {$wpdb->siteid} AND spam = '0' AND deleted = '0' AND archived = '0' AND blog_id > {$max_blog_id} ORDER BY blog_id ASC LIMIT {$offset}" );

  if ( ! $site_ids ) {
    break;
  }

  foreach ( (array) $site_ids as $site_id ) {
    // do your thing
  }

  $max_blog_id = end( $site_ids );
}
```